### PR TITLE
[FIX-3177]Task time parameter parsing error

### PR DESCRIPTION
**Pull request by CalvinKirs**

### CODE REVIEW

**Review by lenboo — APPROVED** [review]
+1

**Comment by lenboo** [comment]
link the issue (#3177)

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/ParameterUtils.java [hunk 1]
```diff
@@ -66,11 +66,13 @@ public static String convertParameterPlaceholders(String parameterString, Map<St
         }
         Date cronTime;
         if (parameterMap != null && !parameterMap.isEmpty()) {
+            // replace variable ${} form,refers to the replacement of system variables and custom variables
+            parameterString = PlaceholderUtils.replacePlaceholders(parameterString, parameterMap, true);
+        }
+        if (parameterMap != null && null != parameterMap.get(Constants.PARAMETER_DATETIME)) {
```
Comment (zhuangchong) [inline dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/ParameterUtils.java:72]: I think you should make sure that the parameters within the Dateutils.parse method are not null

```
if (StringUtils.isBlank(date) || StringUtils.isBlank(format)) {
            return null;
        }
```

```
 public static String convertParameterPlaceholders(String parameterString, Map<String, String> parameterMap) {
        if (StringUtils.isEmpty(parameterString)) {
            return parameterString;
        }

        if (parameterMap == null || parameterMap.isEmpty()) {
            return dateTemplateParse(parameterString, new Date());
        }

        //Get current time, schedule execute time
        String cronTimeStr = parameterMap.get(Constants.PARAMETER_DATETIME);
        Date cronTime = DateUtils.parse(cronTimeStr, Constants.PARAMETER_FORMAT_TIME);
        if (cronTime == null) {
            return parameterString;
        }
        // replace variable ${} form,refers to the replacement of system variables and custom variables
        parameterString = PlaceholderUtils.replacePlaceholders(parameterString, parameterMap, true);
        return dateTemplateParse(parameterString, cronTime);
    }
```

Comment (CalvinKirs) [reply dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/ParameterUtils.java:72]: I think this is not necessary, if it is null, then no replacement is performed.
In this case, the time itself is wrong. So it doesn’t make sense to replace or not

**Comment by CalvinKirs** [comment]
image
Sonar cannot count the coverage of powermock. When preForTest contains test classes.

**Comment by xingchun-chen** [comment]
+1

**Pull request merged**